services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/ApplicationService.kt

```diff
@@ -248,23 +250,32 @@ open class ApplicationService(
      * delete, if the application is in Allu, and it's beyond the pending status.
      */
     @Transactional
-    open fun delete(id: Long, userId: String) {
-        val application = getById(id)
-        val alluid = application.alluid
-        logger.info { "Deleting application, id=$id, alluid=$alluid userid=$userId" }
+    open fun delete(applicationId: Long, userId: String) =
+        with(getById(applicationId)) {
+            cancelIfSent(alluid, id)
+            deleteAndLog(this, userId)
+        }
 
-        if (alluid == null) {
-            logger.info { "Application not sent to Allu yet, simply deleting it. id=$id" }
-        } else {
-            logger.info {
-                "Application is sent to Allu, trying to cancel it before deleting. id=$id alluid=$alluid"
+    /**
+     * Deletes an application. Cancels the application in Allu if it's still pending. Refuses to
+     * delete, if the application is in Allu, and it's beyond the pending status.
+     *
+     * Furthermore, if the owning Hanke is generated and has no more applications, also deletes the
+     * Hanke.
+     */
+    @Transactional
+    open fun deleteWithOrphanGeneratedHankeRemoval(applicationId: Long, userId: String) =
+        with(getById(applicationId)) {
+            cancelIfSent(alluid, id)
+            deleteAndLog(this, userId)
```

Comment:
The beginning is exactly the same as the whole of delete. Is there a way to call delete here instead?

Reply:
They both need to fetch the application entity. I did not want getById(applicationId) to be called twice.